UefiTestingPkg/Include/Library/FlatPageTableLib.h

```diff
@@ -0,0 +1,166 @@
+  UINT32            ArchSignature;
+  PAGE_MAP_ENTRY    *Entries;
+  UINTN             EntryCount;
+  UINTN             EntryPagesAllocated;
+} PAGE_MAP;
+
+// The signature of the PAGE_MAP struct is used to determine the architecture of the page/translation table
+// entries.
+#define AARCH64_PAGE_MAP_SIGNATURE  SIGNATURE_32 ('A','A','6','4')
+#define X64_PAGE_MAP_SIGNATURE      SIGNATURE_32 ('X','6','4',' ')
+
+// This union can be used to interpret each PAGE_MAP_ENTRY on an AARCH64 system.
+typedef union {
+  struct {
+    UINT64    Valid                      : 1;  // BIT0
+    UINT64    BlockOrTable               : 1;  // BIT1
+    UINT64    AttributeIndex             : 3;  // BIT2-4
+    UINT64    NonSecure                  : 1;  // BIT5
+    UINT64    AccessPermissions          : 2;  // BIT6-7
+    UINT64    Shareability               : 2;  // BIT8-9
+    UINT64    AccessFlag                 : 1;  // BIT10
+    UINT64    NonGlobal                  : 1;  // BIT11
+    UINT64    Oa                         : 4;  // BIT12-15
+    UINT64    Nt                         : 1;  // BIT16
+    UINT64    OutputAddress              : 33; // BIT17-49
+    UINT64    Guarded                    : 1;  // BIT50
+    UINT64    Dirty                      : 1;  // BIT51
+    UINT64    Contiguous                 : 1;  // BIT52
+    UINT64    Pxn                        : 1;  // BIT53
+    UINT64    Xn                         : 1;  // BIT54
+    UINT64    Ignored                    : 4;  // BIT55-58
+    UINT64    PageBasedHardwareAttribute : 4;  // BIT59-62
+    UINT64    Reserved                   : 1;  // BIT63
+  } Bits;
+  UINT64    Uint64;
+} AARCH64_PAGE_MAP_ENTRY;
+
+// This union can be used to interpret each PAGE_MAP_ENTRY on an X64 system.
+typedef union {
+  struct {
+    UINT64    Present              : 1;  // BIT0
+    UINT64    ReadWrite            : 1;  // BIT1
+    UINT64    UserSupervisor       : 1;  // BIT2
+    UINT64    WriteThrough         : 1;  // BIT3
+    UINT64    CacheDisabled        : 1;  // BIT4
+    UINT64    Accessed             : 1;  // BIT5
+    UINT64    Dirty                : 1;  // BIT6
+    UINT64    Pat                  : 1;  // BIT7
+    UINT64    Global               : 1;  // BIT8
+    UINT64    Reserved1            : 3;  // BIT9-11
+    UINT64    PageTableBaseAddress : 40; // BIT12-51
+    UINT64    Reserved2            : 7;  // BIT52-58
+    UINT64    ProtectionKey        : 4;  // BIT59-62
+    UINT64    Nx                   : 1;  // BIT63
+  } Bits;
+  UINT64    Uint64;
+} X64_PAGE_MAP_ENTRY;
+
+// When the page/translation table is parsed to create an array of PAGE_MAP_ENTRY, the following bitmasks
+// are used to determine the attributes of one page/translation table entry are the same as another
+// page/translation table entry. If contiguous leaf/block entries have the same attributes, then they
+// will be represented in a single PAGE_MAP_ENTRY.
+#define AARCH64_ATTRIBUTES_MASK  ((0xFFFULL << 52) | (0x3FFULL << 2))

```

Comment:
are these needed in public api and used by caller?

Do these operate on *_PAGE_MAP_ENTRY or on the arch defined page table entry.

Reply:
These operate on the architecturally defined table entries. They are not needed in the public API and I put them here just to inform the library user how the logic should differentiate determine likeness between two entries in the page table.

The bitmask definitions can just be put here as a comment, though. I'll move the #define into the arch specific .c and leave a comment noting what the bitmasks are.